overlord/snapshotstate/backend/backend.go

```diff
@@ -544,13 +568,72 @@ func (t *importTransaction) Cancel() error {
 
 // Commit will commit a given transaction
 func (t *importTransaction) Commit() error {
-	if err := os.Remove(importInProgressFilepath(t.id)); err != nil {
+	t.committed = true
+	return t.unlock()
+}
+
+func (t *importTransaction) lock() error {
+	flock, err := osutil.NewFileLock(t.flockPath)
+	if err != nil {
 		return err
 	}
-	t.committed = true
+	if err := flock.Lock(); err != nil {
+		return err
+	}
+	t.flock = flock
 	return nil
 }
 
+func (t *importTransaction) unlock() error {
+	if t.flock == nil {
+		return nil
+	}
+	if err := t.flock.Close(); err != nil {
+		return err
+	}
+	if err := os.Remove(t.flockPath); err != nil {
+		return err
+	}
+	t.flock = nil
+	return nil
+}
+
+// CleanupAbandondedImports will clean any import that is in progress.
+// This is meant to be call at startup of snapd before any real imports
+// happen.
+//
+// The amount of snapshots cleaned is returned or an error
+func CleanupAbandondedImports() (cleaned int, err error) {
+	inProgressSnapshots, err := filepath.Glob(filepath.Join(dirs.SnapshotsDir, "*_importing"))
+	if err != nil {
+		return 0, err
+	}
+
+	var errs []error
+	for _, p := range inProgressSnapshots {
+		tr, err := newImportTransactionFromImportFile(p)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+		if tr.InProgress() {
+			continue
+		}
+		if err := tr.Cancel(); err != nil {
+			errs = append(errs, err)
+		}
+		cleaned++
+	}
+	if len(errs) > 0 {
+		buf := bytes.NewBuffer(nil)
+		for _, err := range errs {
+			fmt.Fprintf(buf, " - %v\n", err)
+		}
+		return cleaned, fmt.Errorf("cannot cleanup imports:\n%s", buf.String())
```

Comment:
Per above comments this may succeed with some cleanups and fail on others, so maybe "cannot cleanup imports:" if cleaned == len(inProgressSnapshots) and "cannot cleanup some imports:" otherwise?